### PR TITLE
Review

### PR DESCRIPTION
**Pull request by derElektrobesen**

### CODE REVIEW

File: server_test.go [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+	s.clockMock.Add(31 * time.Second) // 1m1s elapsed, nextBackoff == 1m
+
+	s.ass.Equal(time.Duration(0), s.s.retryTimeout())
+
+	// Test timeouts
+	ready := make(chan struct{})
+	ctx, cancel := context.WithCancel(ctx)
+	s.dialerMock.EXPECT().
+		Dial(gomock.Any(), gomock.Any()).
+		DoAndReturn(func(ctx context.Context, _ string) (net.Conn, error) {
```
Comment (golangci) [inline server_test.go]: testServerIsDown$1 - ctx is unused

File: config.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
+package goconnpool
+
+import "time"
+
+const (
+	// DefMaxConnsPerServer is the default value for MaxConnsPerServer config variable.
+	DefMaxConnsPerServer = 1
```
Comment (shmel1k) [inline config.go]: Its better to name as 'DefaultMax...'
see http.DefaultClient for example

File: config.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	DefMaxBackoffInterval = 30 * time.Second
+)
+
+// Config holds some fields required during new connection establishing.
+type Config struct {
+	// MaxConnsPerServer declares maximum number of opened connections per one registered server.
+	// Default value is DefMaxConnsPerServer.
+	MaxConnsPerServer int
+
+	// MaxRPS declares muximum number of requests which could be sent into one server (XXX: not connection)
```
Comment (shmel1k) [inline config.go]: maximum :)

File: server_test.go [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+	s.clockMock.Add(31 * time.Second) // 1m1s elapsed, nextBackoff == 1m
+
+	s.ass.Equal(time.Duration(0), s.s.retryTimeout())
+
+	// Test timeouts
+	ready := make(chan struct{})
+	ctx, cancel := context.WithCancel(ctx)
+	s.dialerMock.EXPECT().
+		Dial(gomock.Any(), gomock.Any()).
+		DoAndReturn(func(_ context.Context, _ string) (net.Conn, error) {
```
Comment (golangci) [inline server_test.go]: testServerIsDown$1 - result 1 (error) is always nil

File: connpool.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+package goconnpool
+
+import (
+	"context"
+	"fmt"
+	"net"
+)
+
+var (
+	errRatelimited  = fmt.Errorf("ratelimited")
```
Comment (shmel1k) [inline connpool.go]: mb its better to name 'ratelimit'?

File: connpool.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+type ConnPool interface {
+	// OpenConnNonBlock requests one connection from the pool.
+	//
+	// If the pool already contains opened connection, this connection will be returned.
+	//
+	// If each registered server is down, function returns an error.
+	// Otherwise function returns active connection (to some alive server in round-robin order)
+	// which could be used to send any type of request.
+	//
+	// Connection should be closed after usage. This action will return connection into pool.
```
Comment (shmel1k) [inline connpool.go]: Strange comment or I dont understand smth.
Why connection is going to be 'closed'? Does it mean that you open connection, use it and close after all? Or you should just return connection to the pool without really closing this connection?

File: containers_test.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestDeck(t *testing.T) {
+	t.Parallel()
+	ass := require.New(t)
```
Comment (shmel1k) [inline containers_test.go:11]: ass
ahahahah 👍

File: pool_impl.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+	cfg = cfg.withDefaults()
+
+	return &connPool{
+		cfg:                 cfg,
+		connProviderFactory: newServerWrapper, // required for tests
+	}
+}
+
+func (p *connPool) OpenConn(ctx context.Context) (Conn, error) {
+	for {
```
Comment (shmel1k) [inline pool_impl.go:35]: Forever till !ctx.Done()? Don't you want to use some timeout, when everything really drops or some 'redialForever' option?

File: pool_test.go [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+	s.cfg.Clock.After(0)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel() // to prevent locking
+
+	s.cfg.Dialer.Dial(ctx, "127.0.0.1:95328") // nolint:errcheck
+}
+
+func testDefaultConnPoolCreation(t *testing.T) {
+	// this test is also exisists just to increment code coverage: nothing to test here %)
```
Comment (shmel1k) [inline pool_test.go]: also exists to increase :)

File: server.go [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+		clock: cfg.Clock,
+	}
+}
+
+func (s *server) updateLastUsage() bool {
+	if s.getRatelimitTimeout() > 0 {
+		return false
+	}
+
+	s.lastUsage = s.clock.Now()
```
Comment (shmel1k) [inline server.go:85]: You dont use mutex here because you dont care about real last usage timestamp, yeah?

File: server.go [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+	go func() {
+		cn, err = s.dialer.Dial(ctx, s.addr)
+		close(ready)
+	}()
+
+	select {
+	case <-ctx.Done():
+		// Check connection were really done
+		select {
+		case <-ready:
```
Comment (shmel1k) [inline server.go]: Some corner-case when dial completes 'just in time' as ctx.Done()?

File: server.go [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+			return cn, err
+		default:
+			return nil, errors.New("timeout")
```
Comment (shmel1k) [inline server.go]: Its better to improve error message.

File: server_test.go [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+
+	"github.com/benbjohnson/clock"
+	gomock "github.com/golang/mock/gomock"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+)
+
+type testServer struct {
+	t   *testing.T
+	ass *require.Assertions
```
Comment (shmel1k) [inline server_test.go:19]: ass <3

File: server_test.go [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+	return s
+}
+
+func (s testServer) withoutTimeouts() testServer {
+	s.cfg.ConnectTimeout = 24 * time.Hour
+	return s
+}
+
+func (s testServer) withoutRateLimits() testServer {
+	s.cfg.MaxRPS = 99999999
```
Comment (shmel1k) [inline server_test.go]: math.MaxInt? :)

**Review by shmel1k — CHANGES_REQUESTED** [review]

File: config.go (hunk 2 above)
Comment (derElektrobesen) [reply config.go]: Fixed

File: config.go (hunk 3 above)
Comment (derElektrobesen) [reply config.go]: Fixed

File: connpool.go (hunk 5 above)
Comment (derElektrobesen) [reply connpool.go]: fixed

File: connpool.go (hunk 6 above)
Comment (derElektrobesen) [reply connpool.go]: Fixed. Close at now just closes the connection, ReturnToPool returns the connection to pool %)

File: pool_impl.go (hunk 8 above)
Comment (derElektrobesen) [reply pool_impl.go:35]: For what?
Timeout could be made with context.WithTimeout() wrapper.

File: pool_test.go (hunk 9 above)
Comment (derElektrobesen) [reply pool_test.go]: Thanks. Fixed

File: server.go (hunk 10 above)
Comment (derElektrobesen) [reply server.go:85]: `updateLastUsage` is called from mutex opened in the previous stack-frame: `getConnection` call.

File: server.go (hunk 11 above)
Comment (derElektrobesen) [reply server.go]: Tnx

File: server.go (hunk 12 above)
Comment (derElektrobesen) [reply server.go]: Done

File: server_test.go (hunk 14 above)
Comment (derElektrobesen) [reply server_test.go]: Not important here %)

File: pool_impl.go (hunk 8 above)
Comment (shmel1k) [reply pool_impl.go:35]: hm, okay
mb its better to create helpers like 
```go
func OpenConnWithTimeout(ctx context.Context, tm time.Duration) (Conn, error) { 
ctx1, cancel := context.WithTimeout(ctx, tm)
defer cancel()
return OpenConn(ctx1)
} 
```

but i don't know if they are really required

File: server.go (hunk 10 above)
Comment (shmel1k) [reply server.go:85]: hm, its not really obvious.

File: pool_impl.go (hunk 8 above)
Comment (derElektrobesen) [reply pool_impl.go:35]: hmmm
makes sense

Comment (derElektrobesen) [reply pool_impl.go:35]: Fixed

File: server.go (hunk 10 above)
Comment (derElektrobesen) [reply server.go:85]: I've add a comment =)

**Comment by derElektrobesen** [comment]
Thanks!

**Pull request closed without merging**